### ISSUE TITLE
Unable to clone vm, looks like a timeout

### ISSUE THREAD

**Issue by TheUnixRoot** [Telmate/terraform-provider-proxmox#431]
https://github.com/Telmate/terraform-provider-proxmox/blob/24df605f0a4602fa3f5d231b5770ae68479f8641/proxmox/resource_vm_qemu.go#L783

I'm using Proxmox 6.3-2 and provider 2.9.0

`proxmox_vm_qemu.resource-name: Still creating... [2m30s elapsed]
╷
│ Error: vm locked, could not obtain config
│ 
│   with proxmox_vm_qemu.resource-name,
│   on hosts.tf line 1, in resource "proxmox_vm_qemu" "resource-name":
│    1: resource "proxmox_vm_qemu" "resource-name" {
│ 
╵`

But in proxmox logs I can see:
`transferred: 8306819 bytes remaining: 2178941 bytes total: 10485760 bytes progression: 79.22 %`

**Comment by mleone87** [comment]
adjust resource create timeouts according to your needs

https://www.terraform.io/docs/language/resources/syntax.html#operation-timeouts

**Comment by TheUnixRoot** [comment]
The thing is that it is the API client which is skipping the PM_TIMEOUT variable as well, so it tries to configure before the cloning finish, leading to an error. The resource itself is not throwing a timeout.
Also in 2.8.0 it works as expected.

**Comment by mleone87** [comment]
if you enable debug log from the api we should see the timeout error on the clone operation

PM_TIMEOUT has not been changed recently, in 2.80 it works because there were random waits in general and more specifically a 5s * infinite time wait until the clone ends because the 

`func (c *Client) WaitForCompletion(taskResponse map[string]interface{}) (waitExitStatus string, err error) `

does not work as intended, at lease for very long operation

adding random waits it's not a solution, but I can see how to fix this function with proper go stile waits

**Comment by shdwlkr** [comment]
I have the same issue. 
This happens to a template which has more than one disk on random time. When proxmox finished copying the first disk - I got the "vm locked error".

**Comment by leandroscardua** [comment]
Hi There,

i`m using the latest version 2.9.0, and having the same issue.

**Comment by wOvAN** [comment]
+ resize and clone timedouts

**Comment by TheLinuxGuy** [comment]
I'm seeing the same issue.

**Comment by viper2133** [comment]
This issue is still present in 2.9.1, 2.9.2 and 2.9.3. If a proxmox clone task takes longer than 5 minutes and 20 seconds terraform will try and send the config file which will give an error of: 

Error: vm locked, could not obtain config.

A simple fix is to go back to 2.8.0 which still works.

**Comment by m4dh4t** [comment]
I struggled on this issue for a while, and was able to fix it temporary by tweaking the `pm_timeout` parameter of the provider to a higher value. For 3 `proxmox_vm_qemu` clones, setting it to 600 was enough

**Comment by Syntax3rror404** [comment]
@viper2133 hi, i have the same error with 2.9.1-3. The only chance is to use the 2.8.0 :(

**Comment by viper2133** [comment]
@m4dh4t Thank you! I didn't know there was an option to modify the api time out.

**Comment by cyakimov** [comment]
I'm having this problem too. Tried with the `PM_TIMEOUT` env variable and setting  `pm_timeout` provider parameter. Both gets ignored and terraform bails after ~6 minutes.

**Comment by phillipgreenii** [comment]
i was seeing this with `2.9.10` as well.  I switch back to `2.8.0` and it works.

**Comment by anon8675309** [comment]
Reproduced on v2.9.10.

Tried [adjusting resource create timeouts](https://github.com/Telmate/terraform-provider-proxmox/issues/431#issuecomment-957775206) and it didn't seem to have any effect.  Here are several test runs after changing the resource create timeouts.  I did not change anything at all between runs:

- Test 1: at the 5m50s mark, got a `vm locked, could not obtain config` error
- Test 2: at the 5m mark, `Error: file provisioner error` because the file provisioner could not connect. "timeout - last error: dial tcp 192.168.1.78:22: connect: no route to host"
- Test 3: `Creation complete after 6m21s`
- Test 4: `vm locked, could not obtain config`
- Test 5: the `terraform destroy` thought there was nothing to clean up, resulting in `Error: 500 can't lock file '/var/lock/qemu-server/lock-135.conf' - got timeout` on creation
- Test 6: `Creation complete after 6m17s`
- Test 7: `Creation complete after 6m9s`
- Test 8: `vm locked, could not obtain config`
- Test 9: the `terraform destroy` thought there was nothing to clean up, resulting in `Error: file provisioner error` because the file provisioner could not connect. "timeout - last error: dial tcp 192.168.1.78:22: connect: no route to host"
- Test 10: `Creation complete after 6m28s`

So that gives some concrete data on the different types of errors that can come up and how often they appear. The issue with the `destroy` incorrectly thinking there's nothing to destroy might be a separate issue and if I can dig up more details on that I'll open a new ticket if one does not already exist.  I suspect all the other issues are related to this timeout problem.

People seem to be mentioning setting both `pm_timeout` and `PM_TIMEOUT` to work around this issue.  In case anyone in the future is confused about which is correct environment variable to us, it is [PM_TIMEOUT](https://github.com/Telmate/terraform-provider-proxmox/blob/ff0a5406753af31f96dfb686829b602e4815c2ef/proxmox/provider.go#L109).   It is referred to as the [pm_timeout](https://github.com/Telmate/terraform-provider-proxmox/blob/31c3e5088b15efa5558669930c58df92bccc7cea/docs/index.md#argument-reference) in the documentation (similar to how the `pm_api_url` value is set by the `PM_API_URL` environment variable).

Similar to what others are reporting, I found that setting `PM_TIMEOUT=600` seems to make everything completely stable.  I've redeployed several times in a row without any failures, so this seems like a solid workaround.

In conclusion, I very much look forward to the [fix with proper go style waits[(https://github.com/Telmate/terraform-provider-proxmox/issues/431#issuecomment-957922720)!

**Comment by Marker689** [comment]
PVE 7.2-7; Plugin v2.9.11
cloud-init image size is 2252M. 

Got timeout everytime when pm_parallel > 3:
```
create full clone of drive ide2 (zfsimage:vm-9000-cloudinit)
trying to acquire lock...
 OK
create full clone of drive scsi0 (zfsimage:base-9000-disk-0)
trying to acquire lock...
trying to acquire lock...
 OK
TASK ERROR: clone failed: can't lock file '/var/lock/pve-manager/pve-storage-zfsimage' - got timeout
```
`export PM_TIMEOUT` and `pm_timeout` seems doesn't make effect.

**Comment by GodAtum360** [comment]
FYI I'm still seeing the same problem with 2.9.11

**Comment by kumorikarasu** [comment]
I seem to be getting this when creating multiple machines in parallel. As soon as the first one finishes, the others error out with this same error. They are all being cloned from the same base image, this doesn't seem to be a timeout issue as the time taken it always how long it takes to clone the first machine.

**Comment by Puben** [comment]
I also experience this issue in 2.9.11. 

Currently my workarounds are to reduce the `pm_parallel` value from 4 to 2 (when cloning to a ZFS disk) or change storage disk from ZFS to LVM (here I can increase 'pm_parallel` to 5 (total amount of machines in Terraform plan).

I wonder whether the problem is Proxmox or the Provider?

**Comment by mleone87** [comment]
guys please keep in mind the actual capacity in terms of I/O and CPU of you servers, cloning issues often depends of that

**Comment by AlexKastrytski** [comment]
Hi, i have some issue. Use ZFS. Proxmox 8.0.4 (latest updates) + Proxmox provider.
__________________________________________________________________________________________

()
create full clone of drive ide2 (big2T:vm-9012-cloudinit)
trying to acquire lock...
 OK
create full clone of drive virtio0 (big2T:base-9000-disk-0/vm-9012-disk-0)
trying to acquire lock...
trying to acquire lock...
can't lock file '/var/lock/pve-manager/pve-storage-big2T' - got timeout
TASK ERROR: clone failed: can't lock file '/var/lock/pve-manager/pve-storage-big2T' - got timeout
